pkg/console/complete.go

```diff
@@ -0,0 +1,245 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+package console
+
+import "strings"
+
+// CompleteLine returns a set of candidate TSQL keywords to complete the current input line
+func CompleteLine(line string) []string {
+	idx := strings.LastIndexAny(line, " ;") + 1
+	// we don't try to complete without a starting letter
+	if idx == len(line) {
+		return []string{}
+	}
+	prefix := strings.ToUpper(string(line[idx:]))
+	left := 0
+	right := len(keywords) - 1
+	for left <= right {
+		mid := (left + right) / 2
+		comp := 0
+		if len(keywords[mid]) >= len(prefix) {
+			comp = strings.Compare(prefix, string(keywords[mid][:len(prefix)]))
+		} else {
+			comp = strings.Compare(prefix, keywords[mid])
+		}
+		if comp < 0 {
+			right = mid - 1
+		} else if comp > 0 {
+			left = mid + 1
+		} else {
+			// look up and down the list from mid and return the slice of matching words
+			first := mid - 1
+			last := mid + 1
+			for first >= 0 && strings.HasPrefix(keywords[first], prefix) {
+				first--
+			}
+			for last < len(keywords) && strings.HasPrefix(keywords[last], prefix) {
+				last++
+			}
+			lines := make([]string, last-first-1)
+			for i, w := range keywords[first+1 : last] {
+				lines[i] = mergeLine(line, w, idx)
+			}
+			return lines
+		}
+	}
+	return []string{}
+}
+
+// mergeline appends keyword to line starting at index idx
+// It matches the case of the current character in the line
+func mergeLine(line string, keyword string, idx int) string {
+	upcase := line[idx] >= 'A' && line[idx] <= 'Z'
+	b := strings.Builder{}
+	b.Write([]byte(line[:idx]))
+	if !upcase {
+		b.WriteString(strings.ToLower(keyword))
+	} else {
+		b.WriteString(keyword)
+	}
+	return b.String()
+}
+
+var keywords = []string{
```

Comment:
Eventually we could enable users to override the keyword list with their own file.  We could also add auto complete for commands.

Reply:
a common format for keywords is the plist file - would be interesting for them to be imported from there.
we might even be able to reuse https://github.com/microsoft/vscode-mssql/blob/main/syntaxes/SQL.plist